core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala

```diff
@@ -64,7 +64,8 @@ private[spark] class ShuffleWriteProcessor extends Serializable with Logging {
         // The map task only takes care of converting the shuffle data file into multiple
         // block push requests. It delegates pushing the blocks to a different thread-pool -
         // ShuffleBlockPusher.BLOCK_PUSHER_POOL.
-        if (Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf) && dep.getMergerLocs.nonEmpty) {
+        if (Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf) && dep.getMergerLocs.nonEmpty &&
+            !dep.shuffleMergeFinalized) {
```

Comment:
What happens if the push isn't finalized in the last stage attempt but the task itself is from a new stage attempt?

Reply:
Then the shuffle dependency won't be `finalized` therefore the new task from a new stage attempt would be able to push.

Reply:
Remove `Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf)` ?

Reply:
> Then the shuffle dependency won't be finalized therefore the new task from a new stage attempt would be able to push.

@venkata91  So, would there be data correctness issue on the merger side (ESS)? Who's in charge of the data file maintains on the merger side in the case of the stage rerun? (I think this's a different issue with indeterministic stage return)

Reply:
sgtm